### ISSUE TITLE
Permission denied creating secrets in GCP secret manager

### ISSUE THREAD

**Issue by lallea** [pulumi/pulumi#3944]
When attempting to manage secrets with pulumi_gcp.secretmanager.Secret (python), i get 403 when pulumi attempts to create the secret. I can successfully create other GCP resources in the same stack, including the project where I attempt to create the secret, so I am authenticated. Service secretmanager.googleapis.com is enabled on project. I am able to create a secret using gcloud command line or curl command, so I have appropriate permissions.

Looking at traces and comparing to curl command, I see that the request sent by pulumi/terraform lacks headers "authorization" and "x-goog-user-project", which might hint at the problem.

Secret config:

`pulumi_gcp.secretmanager.Secret(
        "my-secret",
        project="my-project",
        replication={"automatic": True},
        secret_id="my-secret-name")
`

Also mentioned on https://pulumi-community.slack.com/archives/CRFUR2DGB/p1582127424076300

Successful curl command:

`> curl "https://secretmanager.googleapis.com/v1beta1/projects/my-project-88b03ea/secrets?secretId=test-secret" --request "POST"  --header "authorization: Bearer $(gcloud auth print-access-token)"  --header "content-type: application/json" --header "x-goog-user-project: my-project-88b03ea" --data "{\"replication\": {\"automatic\": {}}}"`

Snippet from trace with --debug and TF_LOG=trace:

`debug: Google API Request Details:
    debug: ---[ REQUEST ]---------------------------------------
    debug: incoming output value translated:  -> 
    debug: POST /v1beta1/projects/my-project-88b03ea/secrets?alt=json&secretId=my-secret-name HTTP/1.1
    debug: incoming output property translated: id -> id
    debug: Host: secretmanager.googleapis.com
    debug: User-Agent: HashiCorp Terraform/0.11+compatible (+https://www.terraform.io) Terraform Plugin SDK/1.4.0 terraform-provider-google-beta/dev
    debug: incoming output value translated: projects/my-project -> projects/my-project
    debug: Content-Length: 33
    debug: Content-Type: application/json
    debug: incoming output property translated: number -> number
    debug: Accept-Encoding: gzip
    debug: 
    debug: incoming output value translated: 608582423015 -> 608582423015
    debug: {
    debug: incoming output property translated: labels -> labels
    debug:  "replication": {
    debug:   "automatic": {}
    debug: incoming output value translated: {} -> {}
    debug: incoming output property translated: projectId -> project_id
    debug:  }
    debug: }
    debug: incoming output value translated: my-project -> my-project
    `

Response:
`    debug: {
    debug:   "error": {
    debug:     "code": 403,
    debug:     "message": "Permission denied on resource project my-project-88b03ea.",
    debug:     "status": "PERMISSION_DENIED",
    debug:     "details": [
    debug:       {
    debug:         "@type": "type.googleapis.com/google.rpc.Help",
    debug:         "links": [
    debug:           {
    debug:             "description": "Google developer console API key",
    debug:             "url": "https://console.developers.google.com/project/my-project-88b03ea/apiui/credential"
    debug:           }
    debug:         ]
    debug:       }
    debug:     ]
    debug:   }
    debug: }
`

**Comment by leezen** [comment]
I was able to create a secret via the call you provided above. Are you logged in via `gcloud auth application-default login` per https://www.pulumi.com/docs/intro/cloud-providers/gcp/setup/ or are you using some other method? Is `gcp:project` set in your stack config?

**Comment by stack72** [comment]
Folks, this 100% works - I have just tested this (due to writing a blog post about it) and followed the following steps:

1. create new project (pulumi new gcp-python)

2. gcloud auth login

3. wrote the code below

```python
import pulumi
from pulumi_gcp import secretmanager

config = pulumi.Config()
my_secret = config.get_secret("my-secret")

secret = secretmanager.Secret("my-secret",
				secret_id="demo-secret",
				replication={
					"automatic": "true",
				})

secret_version = secretmanager.SecretVersion("my-secret-version-1",
						secret=secret.id,
						secret_data=my_secret)
```

4. `pulumi config set my-secret Password1234!`

5. `pulumi up`

```
▶ pulumi up
Previewing update (dev):
     Type                                Name                 Plan
 +   pulumi:pulumi:Stack                 secret-testing-dev   create
 +   ├─ gcp:secretmanager:Secret         my-secret            create
 +   └─ gcp:secretmanager:SecretVersion  my-secret-version-1  create

Resources:
    + 3 to create

Do you want to perform this update? yes
Updating (dev):
     Type                                Name                 Status
 +   pulumi:pulumi:Stack                 secret-testing-dev   created
 +   ├─ gcp:secretmanager:Secret         my-secret            created
 +   └─ gcp:secretmanager:SecretVersion  my-secret-version-1  created

Resources:
    + 3 created

Duration: 9s

Permalink: https://app.pulumi.com/stack72/secret-testing/dev/updates/1
```

**Comment by leezen** [comment]
Thanks double-checking @stack72 
Closing this for now since we haven't heard back from the issue creator

**Comment by lallea** [comment]
Thanks for looking at it. It failed for me in a stack where other GCP operations succeeded, so something was weird.

I haven't had time to make a smaller reproduction yet. I'll reopen if I manage to reproduce.

**Comment by stack72** [comment]
Thanks @lallea for sure ping me here if it continues to be a problem

**Comment by lallea** [comment]
I can confirm that this works now. I am now running pulumi 1.11.1, and presume that the problem disappeared with an upgrade from an older version.

**Comment by stack72** [comment]
Nice one! Thanks for circling back @lallea

**Comment by lallea** [comment]
It wasn't the upgrade. I managed to trigger it again now that I had a working example, and figured it out. 

I used the project parameter, and referred to a project resource in the same stack (which I didn't write above...). If I try `Secret(..., project=my_project.name, ...)`, the project parameter resolves to `my-project-7890abcd` or similar, and I get a 403. If I use `project=my_project.id`, it resolves to `projects/my-project`, and I get a 404. Same if I try `project=my_project`. So I have to build the string argument to `project` by hand: `project="my-project"`.

I think this is a symptom of stringly typing. In an API with typed resources, it would have been natural to detect the problem early and provide a relevant error message. I think that Pulumi is on to something great, so I hope for a typed interface in some distant future. :-)

**Comment by johnpena** [comment]
I also encountered this same issue - able to create secrets/versions manually and with the gcloud command line tool, but got a 403 from `pulumi up`. I was able to fix it by doing:

```
$ gcloud auth application-default login
```

and then setting the `GOOGLE_CREDENTIALS` environment variable to the path for the credentials file (often something like `~/.config/gcloud/application_default_credentials.json`).
